src/map/lua/lua_baseentity.cpp

```diff
@@ -683,7 +683,7 @@ inline int32 CLuaBaseEntity::getLastOnline(lua_State* L)
     TPZ_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
     TPZ_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
 
-    if (auto PChar = static_cast<CCharEntity*>(m_PBaseEntity); PChar->objtype == TYPE_PC)
+    if (auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity); PChar && PChar->objtype == TYPE_PC)
```

Comment:
Almost, after `dynamic_cast`, these checks are redundant; as they're implied by `dynamic_cast`: `PChar && PChar->objtype == TYPE_PC`

All you need is:
```cpp
if (auto* PChar = dynamic_cast<CCharEntity*>(m_PBaseEntity))
{
...
```

Reply:
Ah, just saw your comment:

> TYPE_PC check needs to stay, since lua_baseentities could be other types and only PCs have valid online times.

`dynamic_cast<CCharEntity*>` will only return a 'usable' pointer if the target has vtable entries and memory allocated for the type it's trying to cast to, otherwise it will fail. Like you mention; `static_cast` will just smush the pointer into the new type and hope for the best. It would have the base class available, but would fail when it tries to get anything from `CharEntity`.

Reply:
Good to know